src/hotspot/share/utilities/vmError.cpp

```diff
@@ -166,6 +170,48 @@ static void print_bug_submit_message(outputStream *out, Thread *thread) {
   out->print_raw_cr("#");
 }
 
+static bool check_stack_headroom(Thread* thread,
+                                 size_t headroom) {
+  static const address stack_top    = thread != nullptr
+                                    ? thread->stack_base()
+                                    : os::current_stack_base();
+  static const size_t  stack_size   = thread != nullptr
+                                    ? thread->stack_size()
+                                    : os::current_stack_size();
+  static const address stack_bottom = stack_top - stack_size;
+
+  const address stack_pointer = os::current_stack_pointer();
+
+  const ptrdiff_t stack_headroom = stack_pointer - stack_bottom;
+  return (stack_pointer < stack_bottom || stack_headroom < 0 ||
+          static_cast<size_t>(stack_headroom) < headroom);
```

Comment:
Could be shortened. E.g. `return stack_pointer - headroom < stack_bottom` ?

Reply:
I reworked the logic. It should be clearer now. And deal properly with over-/underflows. 
I noticed when testing on different OS that not including the stack guard size was a bad idea. Some platforms had more than 64K as a guard size.